### PR TITLE
Custom UI: extract device token from micloud

### PR DESCRIPTION
**Pull request by nVuln**
I make a custom form to extract device token inside plugin setting page, it's all automatically

Screen Shot 2021-08-23 at 3 00 17 PM

also, added condition for fieldset in config schema, made it simplest as much as possible

Screen Shot 2021-08-23 at 3 03 11 PM

### CODE REVIEW

**Comment by merdok** [comment]
Very nice👍 Thanks for the pull request! Since I am currently on holiday, I will review this next week and merge it then!

File: lib/protocol/MiCloud.js [hunk 1]
```diff
@@ -2,6 +2,7 @@ const crypto = require('crypto');
 const fetch = require('node-fetch');
 const randomstring = require('randomstring');
 const querystring = require('querystring');
+const TwoFactorRequired = require("../errors/TwoFactorRequired");
```
Comment (merdok) [inline lib/protocol/MiCloud.js:5]: Can you elaborate why this is needed?

File: lib/errors/TwoFactorRequired.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module.exports = class TwoFactorRequired extends Error {
+  constructor(notificationUrl) {
+    super('Two factor authentication required, please use visit below url and retry login. ' + notificationUrl);
+
+    this.notificationUrl = notificationUrl;
+  }
+}
```
Comment (merdok) [inline lib/errors/TwoFactorRequired.js:7]: Also why this file is needed?

**Comment by merdok** [comment]
Hi, 
i have added a couple of comments. Could you please explain what is the `notificationUrl ` in the MiCloud protocol? 
Also why the TwoFactorRequired error was placed in a separate module?
Since i am not familiar with homerbidge-ui yet, are all the files (server, jquery, bootstrap) needed for it to run?

File: lib/protocol/MiCloud.js [hunk 3]
```diff
@@ -318,9 +319,14 @@ class MiCloud {
...
       userId,
-      location
+      location,
+      notificationUrl,
     } = this._parseJson(content);
 
+    if (notificationUrl) {
+      throw new TwoFactorRequired(notificationUrl);
+    }
+
```
[the remark this answers is not in the document]
Comment (nVuln) [reply lib/protocol/MiCloud.js:329]: sometime, MiCloud asking people to verify their connection before login, people need visit notificationUrl on their browser to verify it, and re-process
look like in your main process have using MiCloud also, so I added this exception, maybe you will need it
you can look into this library: https://github.com/PiotrMachowski/Xiaomi-cloud-tokens-extractor/commit/53643cc201ec9a2f598dd739a4bbae478fcb8b2b , they're handled this also

**Comment by nVuln** [comment]
> Since i am not familiar with homerbidge-ui yet, are all the files (popper.js, jquery.js, bootstrap.js) needed for it to run?

popper.js, jquery.js, bootstrap.js is just frontend library support, used in index.html only, currently I used it for collapsed effect in UI and some event listener, maybe more feature in future
if you don't want it in your code, I can remove it and convert exists code to plain javascript

**Comment by merdok** [comment]
Thanks for explaining! Understood! Great work and thanks for the contribution!

**Pull request merged**